ADR/0028-component-dependencies.md

```diff
@@ -0,0 +1,179 @@
+* We will call a Component on which other Components have a dependency: a **"dependency"** component.
+* [integration-service] will handle the testing and promotion of all **dependancy** components in a special way.
+* [build-service] will use the `depends on` links to propose updates to users' **dependant** git repositories.
+* The set of Components and their "depends on" links form a directed-acyclic-graph (DAG). Application-service is responsible for making sure that cycles are not introduced and for managing references in the lifecycle of all Components in the Application. Deleting a Component should be well-defined.
+* As with Components in general, the "depends on" references are Application-scoped. Declared relationships between Components in different Applications are not to be considered valid. Declared relationships between Components in different namespaces are not to be considered valid.
+
+**PR Groups**: Introduce a new convention for PRs that lets one PR declare that it is related to another PR.
+
+* Treat the **source branch name** of the PR as a slash-delimited string, and use the first element of that
+  list as the PR Group name.
+* A new set of functionality for [integration-service] uses PR Groups to guide testing of related
+  PRs.
+* The set of PRs in a PR Group form a flat set.
+* PR Groups are scoped to Component repos under a single Application. Even if PRs across different
+  applications share the same **source branch name**, they shouldn't be considered part of the
+  same PR Group.
+* For example, if the source branch name for a PR is `my-feature/change-1`, then the PR Group
+  name should be interpreted as `my-feature`.
+* A source branch name with no slashes is the degenerate case. For example, if the source branch
+  name for a PR is `my-feature`, then the PR Group name should be interpreted as `my-feature`.
+* Since all PRs have a source branch name, all PRs are trivially members of a PR Group.
+
+Important to understand: the declared dependencies as defined on the Components' `depends on` lists and PR Groups are two different things:
+
+* Component dependencies are declarations about which Components' builds need each other. Automation uses this to create PRs in PR Groups.
+  * You say in english: “Builds of Component A depend on builds of Component B.”
+* PR Groups can be supplied by the user without any declared Component dependencies:
+  * You say in english: “PR #1 and PR #2 are in the same PR Group.”
+
+**Context directories** for [Pipelines as Code](https://pipelinesascode.com/)(PaC): introduce a new `spec.context` field in the `Repository` CR. When present, this field instructs PaC to only create a PipelineRun for events which include changes to the declared context directory in the git repository. This enables us to only run PipelineRuns that are relevant to Components that change in a given push or pull request event. By default, this is ".": the entire git repository.
```

Comment:
Since each repo can have only one Repository CR, this will break the support for mono repos.

Reply:
Even without this limitation I struggle to see how this component approach would work in the real world with a monorepo. I think it would end up being really confusing because each 'component' subdir would essentially be versioned separately, based on which files were touched in a given commit.

Reply:
I agree with the potential for confusion here. On one level, it may not be confusing because all references will be uniquely recorded in the Snapshot so there is a definitive source, but on another level, having builds from multiple commits in the same repo within a Snapshot could be confusing because of the inconsistency. This would be additionally confusing if AppStudio ever gets into the business of versioning (i.e. tagging) component releases as it could result in multiple tags being attached to successive commits.

At some level, the use of context directories within the ADR is an implementation detail that can change. We need to create a process which enables multiple components to be built from the same repository that can reference each other. This ADR _assumed_ that the links would be propagated by immutable references (i.e. image digests), but another approach that I discussed in an [earlier comment](https://github.com/redhat-appstudio/book/pull/113#discussion_r1293785944) is to enable "dynamic" pinning on the _same_ reference by modifying when the PipelineRuns are scheduled. If we go this route, we would likely have greater recourse from within EC to _ensure_ that the observed behavior is desired.

Reply:
I think that this approach makes sense over the use of context directories. The pipeline definitions leverage the `{{revision}}` variable from PaC, i.e.
```
[...]
metadata:
  annotations:
    build.appstudio.redhat.com/commit_sha: '{{revision}}'
[...]
spec:
  params:
  - name: output-image
    value: quay.io/organization/workspace/application/component:{{revision}}
```

Therefore the current pipelines would be able to handle this situation to map between "dynamic" tags assuming that there is a way for the builds themselves to modify the dependency references used based on an input (i.e. leveraging `ARGS` to set the revision). Since there may be changes in the context directory during the build, any user build process that needs to base itself off that revision should leverage `git rev-parse HEAD` to ensure a consistent reference is retrieved.

Reply:
To gal's original question "Since each repo can have only one Repository CR, this will break the support for mono repos.", I revised this today in 6d6b36c.